src/cdk/lib/pipeline.ts

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
+import { BuildSpec, LinuxBuildImage } from 'aws-cdk-lib/aws-codebuild';
+import { PipelineType } from 'aws-cdk-lib/aws-codepipeline';
+import { IRole, ManagedPolicy, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
+import { BlockPublicAccess, Bucket, BucketEncryption } from 'aws-cdk-lib/aws-s3';
+import { CodeBuildStep, CodePipeline, CodePipelineSource } from 'aws-cdk-lib/pipelines';
+import { NagSuppressions } from 'cdk-nag';
+import { Construct } from 'constructs';
+import { CoreStage, CoreStageProperties } from './stages/core';
+import { Utilities } from './utils/utilities';
+import { RetentionDays } from 'aws-cdk-lib/aws-logs';
+import { S3Trigger } from 'aws-cdk-lib/aws-codepipeline-actions';
+import { ContainerDefinition, ContainersPipelineStage } from './stages/containers';
+import { StorageStage } from './stages/storage';
+import { AuroraPostgresEngineVersion } from 'aws-cdk-lib/aws-rds';
+
+export interface CDKPipelineProperties extends StackProps {
+    configBucketName: string;
+    branchName: string;
+    organizationName: string;
+    repositoryName: string;
+    workingFolder: string;
+    tags?: { [key: string]: string };
+    coreStageProperties?: CoreStageProperties;
+    defaultRetentionPeriod?: RetentionDays;
+    applicationList: ContainerDefinition[];
+    petImagesPaths: string[];
+    postgresEngineVersion?: AuroraPostgresEngineVersion;
+}
+
+export class CDKPipeline extends Stack {
+    /**
+     * The IAM role used by the pipeline.
+     *
+     * @readonly
+     * @type {IRole}
+     */
+    readonly pipelineRole: IRole;
+    constructor(scope: Construct, id: string, properties: CDKPipelineProperties) {
+        super(scope, id, properties);
+
+        // Create a CodePipeline source using the Specified S3 Bucket
+        const configBucket = Bucket.fromBucketName(this, 'ConfigBucket', properties.configBucketName);
+        const bucketKey = `repo/refs/heads/${properties.branchName}/repo.zip`;
+
+        // Use the configuration file as the pipeline trigger
+        const bucketSource = CodePipelineSource.s3(configBucket, bucketKey, {
+            trigger: S3Trigger.POLL,
+        });
+        /**
+         * Create an S3 bucket to store the pipeline artifacts.
+         * The bucket has encryption at rest using a CMK and enforces encryption in transit.
+         * Versioning is enabled on the bucket for audit and recovery purposes.
+         */
+        const pipelineArtifactBucket = new Bucket(this, 'ArtifactBucket', {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The bucket should have public access restricted and blocked to prevent unauthorized access.